Enable Google Cloud APIs using Terraform

mijndert opened the pull request:
## Description

This PR fixes #12. It enables all necessary GCP services with a Terraform resource. Because Terraform can only enable one service at a time I created a loop that iterates over the gcp_services list.

In your guidelines you state that no changes should be merged to the main branch directly, but there's no other branch to merge this into (yet). Please create a new branch if needed. 

## Testing instructions

Check to see whether all services as described in tfvars are properly enabled.

## Checklist for the submitter

 - [x] Documentation updated
 - [ ] ~~Infrastructure overview diagram updated~~

## Checklist for reviewers

- [ ] ~~If `terraform-hisec/` was updated: someone with the "Infra Owner" role has checked the output of `terraform plan`~~

## Definition of done (for team members)

This checklist is for the person who merges this pull request. _Tick check boxes, or strike out irrelevant ones._

 - [ ] ~~If `terraform-hisec/` was updated: someone with the "Infra Owner" role has run `terraform apply`~~
 - [ ] Tested in production

FooBarWidget commented:
Hey @mijndert, thanks for this pull request! The guidelines state that team members should not _push_ to main directly, and that changes should be proposed via pull requests. So what you're doing is fine!

Your code is a good start. But you should specify dependencies between Terraform resources. Terraform applies all resources concurrently, unless you specify dependencies.

For example, `google_dns_managed_zone` doesn't work until the dns.googleapis.com API is enabled. So the former should have a dependency on the latter. If you don't, then it could happen that Terraform creates the DNS zone before the DNS API is enabled, which results in an error.

Similarly, all other resources should depend on the `google_project_services`s (there may be multiple) that enable the APIs that resources need.

Here's an example:

~~~terraform
resource "google_project_service" "dns" {
  project = var.gcloud_project
  service = "dns.googleapis.com"
} 

resource "google_dns_managed_zone" "fullstaqruby_org" {
  name = "fullstaqruby-org"
  depends_on = [google_project_service.dns]  # <-------- added this
  dns_name = "fullstaqruby.org."
  dnssec_config {
    state = "on"
  }
}
~~~

FooBarWidget commented on terraform/services.tf:
```diff
@@ -0,0 +1,8 @@
+  service = var.gcp_services[count.index]
+  # (Optional) If true, services that are enabled and which depend on this service should also be disabled when this service is destroyed.
+  # If false or unset, an error will be generated if any enabled services depend on this service when destroying it.
+  disable_dependent_services = true
```
I think it's more appropriate to not set `disable_dependent_services`. Rather than disabling services, I want Terraform to fail because it likely signals that we've made a mistake.

mijndert replied:
Agreed, that's a better way of doing it. I changed the code accordingly.

mijndert commented:
@FooBarWidget what if we create a null resource with a wait condition? Because enabling the API's sometimes takes a while which in turn might break depends_on. The null resource will then be dependent on all the different `google_project_service`s.

FooBarWidget commented:
I don't understand what you mean by "breaks depends_on". As far as I know, waiting is the entire point of depends_on. In the above example, Terraform should wait with creating a DNS zone, until google_project_service.dns is done.

Can you show with an example what problem you see?

mijndert commented:
I discovered that Terraform just executes the command to enable the API's and doesn't know when the API is actually available. It can take a few seconds for the API to be available thus a resource that depends on said API might fail to create.

FooBarWidget commented:
Oooh that sucks... is there a way to poll for readiness?

mijndert commented:
I have way more AWS experience so I don't know if you can poll these API's. A wait resource might be easier, but a little 
more error-prone.

```
resource "null_resource" "wait-on-apis" {
  provisioner "local-exec" {
    command = "sleep 60"
  }
  depends_on = [ ... ]
}
```

FooBarWidget commented:
A fixed sleep time is not reliable. I'll ask around and see whether there's a reliable solution for this.

FooBarWidget commented:
I did some research. [This is the bug](https://github.com/hashicorp/terraform-provider-google/issues/1393). But they say it's been fixed in [this PR](https://github.com/hashicorp/terraform-provider-google/pull/1524) in 2018. We're using the Google provider v3.16 so the fix should already apply to us.

I ran the following test case....

~~~terraform
provider "google" {
  version = "~> 3.16.0"
  project = "fullstaq-ruby"
}

resource "google_project_service" "demo" {
  service = "cloudkms.googleapis.com"
}

resource "google_kms_key_ring" "demo" {
  depends_on = [google_project_service.demo]
  name     = "demo"
  location = "europe-west4"
}
~~~

...and this test case correctly waits until the Cloud KMS API is enabled, before creating the KMS key ring.

Have you run into a concrete issue? If so, can you post the error message you see?

mijndert commented:
@FooBarWidget Guess I had some outdated intel, learning on the job here. I'm new to GCP but have 10+ years of experience with AWS - which doesn't help me here. ;-)

I updated the PR with the creation of the API resources and added some dependencies.

lmk

FooBarWidget commented on docs/deploy.md:
```diff
@@ -10,13 +10,12 @@ You need Terraform, kubectl and the Google Cloud CLI.
 gcloud auth application-default login
 gcloud auth login
 gcloud config set project fullstaq-ruby
-../enable_gcloud_apis.sh
```
You should delete the enable_gcloud_apis.sh

FooBarWidget commented on terraform/services.tf:
```diff
@@ -0,0 +1,35 @@
+resource "google_project_service" "compute-api" {
```
The static IP addresses depend on this API.

FooBarWidget replied:
I just found out that service accounts (service_accounts.tf) depend on the iam.googleapis.com API.

FooBarWidget commented on terraform/services.tf:
```diff
@@ -0,0 +1,35 @@
+resource "google_project_service" "compute-api" {
+  project = var.gcloud_project
+  service = compute.googleapis.com
+  disable_dependent_services = false
```
No need to set this to false. Just leave it out altogether.

FooBarWidget commented on terraform/services.tf:
```diff
@@ -0,0 +1,35 @@
+resource "google_project_service" "compute-api" {
+  project = var.gcloud_project
+  service = compute.googleapis.com
```
This must be a string, not a literal.

FooBarWidget commented:
Aaaand deployed to production. Thank you very much!

mijndert commented:
Great, happy to help. On to the next one. :-)
